src/index/base.cpp

```diff
@@ -76,7 +78,7 @@ bool BaseIndex::Init()
         SetBestBlockIndex(locator_index);
     }
     m_synced = m_best_block_index.load() == active_chain.Tip();
-    if (!m_synced) {
+    if (!m_synced && fPruneMode) {
```

Comment:
In commit "index: Enable reindex-chainstate with active indexes" (1e598cfb2b3d0e56b1fb301bcf6870f42453c7cc)

This is pretty opaque and could use a comment. I'm also not sure it is better to be looking at `fPruneMode` here, than just checking whether reindexing is happening, when it sounds like reindexing is the real problem, and checking for `fPruneMode` is just a proxy for avoiding the real source of the problem?

Also it is not obvious to me that just because pruning isn't currently enabled doesn't mean pruning wasn't previously enabled and there couldn't be missing blocks worth checking for here.

Reply:
> Also it is not obvious to me that just because pruning isn't currently enabled doesn't mean pruning wasn't previously enabled and there couldn't be missing blocks worth checking for here.

I think that this shouldn't be possible: We check in Init when loading the chainstate (`chainman.LoadBlockIndex()`) that we either have all the blocks from genesis, or allow for pruning - otherwise, we abort with an InitError and never get to the point where the indexes are started.

I replaced the check with the indirect `g_indexes_ready_to_sync` as suggested and added a comment. This way, the pruning check will continue to be executed regardless of pruning status (unless -reindex is specified).

What I don't like is that this would throw a confusing InitError message if it somehow failed on a non-pruning node ("block of the index goes beyond pruned data"). Maybe the check should assert instead if `fPrune==false`?